agent/stats/engine.go

```diff
@@ -163,77 +200,83 @@ func (engine *DockerStatsEngine) removeAll() {
 		}
 		delete(engine.tasksToContainers, task)
 	}
-}
-
-// listContainersAndStartEventHandler adds existing containers to the watch-list
-// and starts the docker event handler.
-func (engine *DockerStatsEngine) listContainersAndStartEventHandler() {
-	// List and add existing containers to the list of containers to watch.
-	err := engine.addExistingContainers()
-	if err != nil {
-		seelog.Warnf("Error listing existing containers, err: %v", err)
-		return
-	}
-}
-
-// addExistingContainers lists existing containers and adds them to the engine.
-func (engine *DockerStatsEngine) addExistingContainers() error {
-	listContainersResponse := engine.client.ListContainers(false, ecsengine.ListContainersTimeout)
-	if listContainersResponse.Error != nil {
-		return listContainersResponse.Error
-	}
 
-	for _, containerID := range listContainersResponse.DockerIDs {
-		engine.addContainer(containerID)
+	for task := range engine.tasksToHealthCheckContainers {
+		delete(engine.tasksToContainers, task)
 	}
-
-	return nil
 }
 
 // addContainer adds a container to the map of containers being watched.
-// It also starts the periodic usage data collection for the container.
-func (engine *DockerStatsEngine) addContainer(dockerID string) {
-	engine.lock.Lock()
-	defer engine.lock.Unlock()
-
+func (engine *DockerStatsEngine) addContainer(dockerID string) *StatsContainer {
 	// Make sure that this container belongs to a task and that the task
 	// is not terminal.
 	task, err := engine.resolver.ResolveTask(dockerID)
 	if err != nil {
 		seelog.Debugf("Could not map container to task, ignoring, err: %v, id: %s", err, dockerID)
-		return
+		return nil
 	}
 
 	if len(task.Arn) == 0 || len(task.Family) == 0 {
 		seelog.Debugf("Task has invalid fields, id: %s", dockerID)
-		return
+		return nil
 	}
 
 	if task.GetKnownStatus().Terminal() {
 		seelog.Debugf("Task is terminal, ignoring, id: %s", dockerID)
-		return
+		return nil
+	}
+
+	seelog.Debugf("Adding container to stats watch list, id: %s, task: %s", dockerID, task.Arn)
+	statsContainer := newStatsContainer(dockerID, engine.client, engine.resolver)
+	engine.tasksToDefinitions[task.Arn] = &taskDefinition{family: task.Family, version: task.Version}
+
+	if !engine.disableMetrics {
+		// Adding container to the map for collecting stats
+		statsContainer = engine.addToStatsContainerMapUnsafe(task.Arn, dockerID, statsContainer, statsMetricsMap)
+	}
+
+	dockerContainer, err := engine.resolver.ResolveContainer(dockerID)
+	if err == nil && dockerContainer.Container.HealthCheckShouldBeReported() {
+		seelog.Debugf("Adding container to stats health check watch list, id: %s, task: %s", dockerID, task.Arn)
+		engine.addToStatsContainerMapUnsafe(task.Arn, dockerID, statsContainer, statsHealthmetricsMap)
+	}
+
+	return statsContainer
+}
+
+func (engine *DockerStatsEngine) addToStatsContainerMapUnsafe(
+	taskARN, containerID string,
+	statsContainer *StatsContainer,
+	mapType statsEngineMapType) *StatsContainer {
```

Comment:
as per offline conversation, this still seems a bit clunky to refer to a map that you have access from an enum. Can you consider using a function pointer here instead, which returns you the appropriate map to manipulate?